### PR TITLE
Inv export

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY

allow an 'export' mode that optimizes inventory for reuse instead of inspection

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

ansible-inventory
##### ANSIBLE VERSION

```
2.5
```

### CODE REVIEW

**Comment by AlanCoding** [comment]
My testing shows that this is a solution for #30877, testing the example inventory reported in the issue.

```
$ ansible-inventory -i issues/gvars/inventory.ini --list --export
{
    "_meta": {
        "hostvars": {
            "host1": {}, 
            "host2": {}
        }
    }, 
    "all": {
        "children": [
            "raleigh", 
            "ungrouped"
        ]
    }, 
    "raleigh": {
        "hosts": [
            "host1", 
            "host2"
        ], 
        "vars": {
            "airport_code": "RDU"
        }
    }, 
    "ungrouped": {}
}
```

The example lives [in this repo](https://github.com/AlanCoding/Ansible-inventory-file-examples/tree/master/issues/gvars), where the file `raleigh` has the airport code variable in it.

```
$ tree issues/gvars/
issues/gvars/
├── group_vars
│   └── raleigh
└── inventory.ini
```

That case is testing variable plugins. This is also returning the group-level variables that are defined in the inventory file/script. Related to the discussion we had in 30877, `ansible-inventory` is _not_ returning group-level variables from the inventory file if the `--export` option is not used, which is the behavior that was discussed.

**Comment by bcoca** [comment]
@AlanCoding I also added a config entry so you can disable though environment/ansible.cfg if you want to avoid the extra `--export` on the cli

I was also thinking of adding switch for vars_plugins, currently we only ship with the one that does `group_vars/` and `host_vars/` but its possible that users have more of them active and might not want them in the export.

ie.  theoretical aws_metadata.py vars plugin that gets extra data from aws for each host dynamically

**Comment by AlanCoding** [comment]
Excellent, this also works for the `all` group, either with the original variable being [defined in the inventory contents](https://github.com/AlanCoding/Ansible-inventory-file-examples/blob/master/scripts/group_vars.py), [or inside of `group_vars/all`](https://github.com/AlanCoding/Ansible-inventory-file-examples/commit/694f94324bb99b59706c0b20f26e8ae24f490fd7).

@bcoca Thank you, that's a good option to have. If this is going out in 2.5, one problem I have been considering is how we can write a usage that won't break if it uses Ansible 2.4. Simply setting the environment variable would seem to do exactly what we need - use the export option if it is available, but not break the usage of versions before then.

File: lib/ansible/cli/inventory.py [hunk 1]
```diff
@@ -83,11 +87,20 @@ def parse(self):
...
+        # graph
         self.parser.add_option("-y", "--yaml", action="store_true", default=False, dest='yaml',
                                help='Use YAML format instead of default JSON, ignored for --graph')
         self.parser.add_option("--vars", action="store_true", default=False, dest='show_vars',
                                help='Add vars to graph display, ignored unless used with --graph')
 
+        # list
+        self.parser.add_option("--export", action="store_true", default=C.INVENTORY_EXPORT, dest='export',
+                               help="When doing an --list, represent in a way that is optimized for export,"
+                                    "not as an accurate representation of how Ansible has processed it")
```
Comment (alikins) [inline lib/ansible/cli/inventory.py:100]: The help blurb is kind of vague. optimized how and for what?

Comment (bcoca) [reply lib/ansible/cli/inventory.py:100]: for export .. as for 'how' .. it seems too long to detail here.

Comment (AlanCoding) [reply lib/ansible/cli/inventory.py:100]: Just to make an attempt at word-smithing

> When doing an --list, represent shared variables inside groups or the inventory,
> which has a smaller memory footprint but is not Ansible's internal representation

Comment (bcoca) [reply lib/ansible/cli/inventory.py:100]: i'll just merge as is for now, let the 'wordsmithing' update it in latter patches so as to not keep this waiting

**Comment by alikins** [comment]
Seems to work for me...

Out of curiosity, I instrumented some deepdiff compares of the various stages of outputs at
https://gist.github.com/alikins/55945db08b57c53f2c7dc99da7c69376


the compares include  
'--list' vs '--list --export'   (json)
'--list --yaml' vs '--list --export --yaml (yaml)
compare of the py data structures returned from json_inventory() and yaml_inventory()
and deepdiffs of the json/yaml results from --export after a loads/dumps cycle

Aside from str->unicode changes, the main dist in '--list --export --yaml' doesn't include a '_meta' key which I believe is expected behavior.

**Comment by AlanCoding** [comment]
From my example:

```
$ export ANSIBLE_INVENTORY_EXPORT=True
$ ansible-inventory -i scripts/group_vars.py --host=foobar
{
    "foovar": "baz"
}
$ unset ANSIBLE_INVENTORY_EXPORT
$ ansible-inventory -i scripts/group_vars.py --host=foobar
{
    "foovar": "baz", 
    "group_var": "group_val", 
    "group_var2": "group_val2"
}
```

I don't use this, it's not particularly important to me. But this doesn't really match with what you've written in the help text for the option.

> --export              When doing an --list, represent...

**Review by nitzmahone — APPROVED** [review]

**Review by AlanCoding — APPROVED** [review]

**Pull request merged**